pkg/i2gw/ingress2gateway.go

```diff
@@ -35,70 +35,135 @@ import (
 	gatewayv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
 )
 
-func Run(printer printers.ResourcePrinter, namespace string, inputFile string) {
+// ProviderConstructorByName is a map of ProviderConstructor functions by a
+// provider name. Different Provider implementations should add their construction
+// func at startup.
+var ProviderConstructorByName = map[ProviderName]ProviderConstructor{}
+
+// ProviderName is a string alias that stores the concrete Provider name.
+type ProviderName string
+
+// ProviderConstructor is a construction function that constructs concrete
+// implementations of the Provider interface.
+type ProviderConstructor func(conf *ProviderConf) Provider
+
+// ProviderConf contains all the configuration required for every concrete
+// Provider implementation.
+type ProviderConf struct{}
+
+// The Provider interface specifies the required functionality which needs to be
+// implemented by every concrete Ingress/Gateway-API provider, in order for it to
+// be used.
+type Provider interface {
+	CustomResourceReader
+	ResourceConverter
+}
+
+type CustomResourceReader interface {
+
+	// ReadResourcesFromCluster reads custom resources associated with
+	// the underlying Provider implementation from the kubernetes cluster.
+	ReadResourcesFromCluster(ctx context.Context, customResources interface{}) error
```

Comment:
would be nice to provide a param here or some other mechanism to enable providers to initialize a client to fetch custom resources with (so common boilerplate doesn't have to be repeated)

could make it part of provider init to add to a scheme a common client is created with, or otherwise

Reply:
The idea was to pass the client using ProviderConf, but it may be better to pass it directly as an argument to the function. Not sure yet about this one.